ssl/statem/statem_lib.c

```diff
@@ -2403,3 +2403,149 @@ int tls13_restore_handshake_digest_for_pha(SSL *s)
     }
     return 1;
 }
+
+#if !defined(OPENSSL_NO_COMP)
+int tls13_construct_compressed_certificate(SSL *s, WPACKET *pkt, CERT_PKEY *cpk,
+                                           uint8_t *context, size_t context_len)
+{
+    WPACKET tmppkt;
+    BUF_MEM *buf = NULL;
+    int ret = 0;
+    size_t length;
+    size_t max_length;
+    COMP_METHOD *method;
+    COMP_CTX* comp = NULL;
+    int comp_len;
+
+    if (s->ext.compress_certificate_rx == TLSEXT_comp_cert_none
+            || (buf = BUF_MEM_new()) == NULL
+            || !WPACKET_init(&tmppkt, buf)) {
+        goto err;
+    }
+
+    /* Use the |tmppkt| for the to-be-compressed data */
```

Comment:
This code is not called when a regular server certificate is being generated. This is for compressed certificates which must compress the certificate; which is something that `tls_construct_server_certificate()` does not do.

Reply:
Yes, this code called only on receiving the Compressed Certificate message. The first part of function duplicates **tls_construct_server_certificate** logic. Here you can generate a Certificate Message by calling **tls_construct_server_certificate** inside **tls13_construct_compressed_certificate** and than compress result of this function. You can just pass ```WPACKET tmppkt;```  to **tls_construct_server_certificate** and get the same result as the first 20 lines of your code.

Reply:
You are proposing using the same buffer for both compression input and output, that is dangerous when the compression could possibly expand the output (which you don't know until after compression occurs); that is not safe.

It is safer to to get the data to be compressed in a separate buffer then compress into the output buffer (which is the buffer for the message).

Reply:
https://github.com/xpahos/openssl/blob/dcdc35357fbb4c81066b8524bc1192f41a5e2e6b/ssl/statem/statem_srvr.c#L3650

I think it will be better to show in the code what I want to say :)